src/profile-logic/marker-data.js

```diff
@@ -243,10 +255,114 @@ export function extractMarkerDataFromName(
   return newMarkers;
 }
 
+/**
+ * This class is a specialized Map for IPC marker data, using the thread ID and
+ * marker index as the key, and storing IPCPairData (which holds the start/end
+ * times for the IPC message.
+ */
+export class IPCMarkerCorrelations {
+  _correlations: Map<string, IPCPairData>;
+
+  constructor() {
+    this._correlations = new Map();
+  }
+
+  _makeKey(tid: number | void, index: number) {
+    if (tid === undefined) {
+      return index.toString();
+    }
+    return `${tid},${index}`;
+  }
+
+  set(tid: number | void, index: number, data: IPCPairData): void {
+    this._correlations.set(this._makeKey(tid, index), data);
+  }
+
+  get(tid: number | void, index: number): ?IPCPairData {
+    return this._correlations.get(this._makeKey(tid, index));
+  }
+}
+
+/**
+ * This function correlates the sender and recipient sides of IPC markers so
+ * that we can share data between the two during profile processing.
+ */
+export function correlateIPCMarkers(threads: Thread[]): IPCMarkerCorrelations {
+  // Create a unique ID constructed from the source PID, destination PID,
+  // message seqno, and message type. Since the seqno is only unique for each
+  // message channel pair, we use the PIDs and message type as a way of
+  // identifying which channel pair generated this message.
+  function makeMarkerID(srcpid, destpid, messageseqno, messagetype) {
+    return `${srcpid},${destpid},${messageseqno},${messagetype}`;
+  }
+
+  // First, construct a mapping of marker IDs to destination markers for
+  // faster lookup.
+  const dstMarkers = new Map();
+  for (const thread of threads) {
+    if (!thread.stringTable.hasString('IPC')) {
+      continue;
+    }
+    for (let index = 0; index < thread.markers.length; index++) {
+      const data = thread.markers.data[index];
+      if (!data || data.type !== 'IPC' || data.direction !== 'receiving') {
+        continue;
+      }
+      const key = makeMarkerID(
+        data.otherPid,
+        thread.pid,
+        data.messageSeqno,
+        data.messageType
+      );
+      if (dstMarkers.has(key)) {
+        console.warn('Duplicate IPC marker found for key', key);
+      } else {
+        dstMarkers.set(key, { tid: thread.tid, index, data });
+      }
+    }
+  }
+
+  // Now iterate over all the source markers and find the corresponding
+  // destination marker (if any).
+  const correlations = new IPCMarkerCorrelations();
+  for (const thread of threads) {
+    if (!thread.stringTable.hasString('IPC')) {
```

Comment:
Maybe provide a quick comment here to add context. Something like:

> Only check for IPC markers if the string table has the string "IPC". This is an easy optimization to avoid looping through all of the markers if we don't have to.